### DISCUSSION TITLE
Sveltekit SSR with realtime pocketbase

### DISCUSSION THREAD

**Discussion by RbnMach** [pocketbase/pocketbase#3363] [Q&A] [answered]
Hello, I'm trying sveltekit SSR with pocketbase and the problem I have is how to apply pocketbase realtime. I leave here a little context of my code, this is my hooks.server.ts

```javascript
//hooks.server.ts
import { Pb } from '$lib/conexion-database/conexion_pocket';
export const handle = async ({ event, resolve }) => {
	event.locals.pb = Pb();
	event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');
	try {
		if (event.locals.pb.authStore.isValid) {
			event.locals.user = structuredClone(await event.locals.pb.collection('cuenta').authRefresh());
		} else {
			event.locals.user = undefined;
		}
	} catch (_) {
		event.locals.pb.authStore.clear();
	}
	const response = await resolve(event);
	const isProd = process.env.NODE_ENV === 'production' ? true : false;
	response.headers.append(
		'set-cookie',
		event.locals.pb.authStore.exportToCookie({ secure: isProd, sameSite: 'Lax', httpOnly: false })
	);
	return response;
};
```
and of course I also defined app.d.ts
```javascript
// app.d.ts
import PocketBase from 'pocketbase';
declare global {
	namespace App {
		interface Locals {
			pb: PocketBase;
			user;
		}
	}
}
export {};
```
This all works great, as well as user logins or CRUD in general on the server side, but my question is how could I implement Pocketbase real-time subscription on the client side since the connection to Pocketbase is on hooks. server.ts and as you know I can only access said connection from the server side.
I tried with the pb_auth token which is saved in my cookie and I was able to use it on the client side to do the subscription in my lean component, something like:
```javascript
// +page.svelte
	const pb = new PocketBase('https://conectpocketbase.com');
	onMount(async () => {
		pb.authStore.loadFromCookie(document.cookie);//get cookie pb_auth
		pb.authStore.onChange(() => {
			document.cookie = pb.authStore.exportToCookie({ httpOnly: false }); 
		});
		pb.collection('items').subscribe('*', function (e) {
			invalidateAll();
		});
		// const records = await pb.collection('items').getFullList({
		// 	sort: '-created'
		// });
		// console.log('result: ', records);
	});
	onDestroy(() => {
		pb.collection('items').unsubscribe('*');
	});
```
and it works well but the problem appears when I have two sessions such as (user1 and user2) open on different computers subscribed to items, the realtime fulfills its function the first time, but what happens later is that it seems that the sessions intersect leaving me at the end (user1 and user1) or (user2 and user2), it is as if one session had replaced the other, I think this is due to how hooks.server.ts generates the tokens because with each request from either the client or the server the token changes.
Another thing I could argue with this problem is that if I don't use pocketbase subscriptions or realtime, the sessions problem persists when (user1 and user2) update their pages at the same time the sessions cross and one session replaces the other leaving something like (user1 and user1) or (user2 and user2), Could you give me some recommendations on how to use Pocketbase in real time with sveltekit SSR as well as if you would recommend anything with sessions or if you can propose another solution, thank you very much in advance.

**Comment by ganigeorgiev** [comment]
There too many missing details to help you debug it.

Are you sure that you have a new server-side SDK instance for each request?

What is the function that `$lib/conexion-database/conexion_pocket` exports?

For the SSR do you use the Node or Deno (or other runtime) SvelteKit adapter? Note that Deno has builtin support for `localStorage` and it is possible that the SDK instance is loading the previous auth state from there.

How are you testing? When you say _"open on different computers"_ do you mean different devices or different tabs/windows? If the latter note that the browser cookie will be shared between all windows/tabs (_unless one of the windows is in Private/Incognito mode_).

If you are still not able to identify the issue, try to isolate the problem and create a **minimal** reproducible repo (_you can also commit your PocketBase `pb_migrations` or export your collections configuration as json from the Admin UI > Settings > Export collections_) and I'll try to have a look.

**Reply by RbnMach** [reply]
Hello, well if I'm sure there is an SDK instance, also $lib/connection-database/conexion_pocket exports the Pocketbase connection, also the adapter I'm using with Sveltekit is Node, and the tests I perform is: on different computers, different browsers and different tabs where one was tab of incognito. In every test I did when I update almost simultaneously the two pages of the two users who logged in, one session replaces the other.

I leave [here](https://github.com/RbnMach/proyect-test-pocketbase.git) the base project where I isolated the problem that occurs to me regarding the sessions, also leave the collection configuration in the same repository in JSON (pb_schema.json).
Thanks in advance for your time...

**Reply by ganigeorgiev** [reply, marked as answer]
In https://github.com/RbnMach/proyect-test-pocketbase/blob/main/src/lib/conexion-database/conexion_pocket.ts, you are returning a global singleton PocketBase instance.

This is OK **only** in browser context or on the server-side if using PocketBase as admin.

But when you want to manage different users requests and their auth states in a long running server-side process, you'll need a unique PocketBase SDK instance for each request (that you can store in `event.locals`). Otherwise requests from different users will write over the same `pb.authStore` instance.

In other words, either change the function in the above file to be a "factory", aka. to return a `new PocketBase()` and not a global variable. Or directly call `new PocketBase()` in your `hooks.server.ts`.

---

As a side-note, you may not need SSR. A lot of devs seems to jump straight to SSR and RSC but you need to remember that once you do that you no longer have "just frontend" and practically you have to maintain 2 different backends, the SvelteKit Node adapter and PocketBase.

I'll try to prepare more detailed client-side recommendations and guides in the next couple weeks, but in general my advice would be if you are not making a website (aka. SEO is not important), then consider client-side rendering whenever possible as it makes things a lot easier and use server-side Node actions only for some one-off special tasks like verifying a payment webhook or some extra form submit postprocessing that can be done only server-side (eg. captcha validation).

**Reply by RbnMach** [reply]
Thank you very much for your help and your recommendations, I will leave the solution applied in this repository here in case someone else might be interested.
BASIC SVELTEKIT SSR (NODE ADAPTER) WITH POCKETBASE CRUD AND REALTIME
[https://github.com/RbnMach/proyect-test-pocketbase](https://github.com/RbnMach/proyect-test-pocketbase)

**Reply by towbeeWee** [reply]
Hi, I'm developing a website where I need the exact functionality of [https://github.com/RbnMach/proyect-test-pocketbase](https://github.com/RbnMach/proyect-test-pocketbase). Unfortunately the realtime function only works if `httpOnly: flase` is set. 
`event.locals.pb.authStore.exportToCookie({ secure: isProd, sameSite: 'Lax', httpOnly: false })`
What solution can I implement so that the realtime function also works with `httpOnly: flase`?
I appreciate every suggestion. thanks in advance

**Reply by ganigeorgiev** [reply]
@towbeeWee 

> Unfortunately the realtime function only works if httpOnly: flase is set.

If you want access to the cookie and to be able to load it in a browser SDK instance then you'll need to export with `httpOnly: false`. Why it is unfortunate? Why do you need JS SSR and mixed PocketBase SDK access in the first place?

**Reply by towbeeWee** [reply]
In my application, users log in on the server side. Same as in the linked repo. I protect some areas serverside using the pb.authStore.model / userid, run webhooks, etc. Serverside makes me feel safer. Maybe I'm thinking wrong?
Unfortunate, because `httpOnly:false` isn't quite as secure than `httpOnly:true`  [>>](https://www.cookiepro.com/knowledge/httponly-cookie/). Do you think I can use `httpOnly:false` without any problem? What is your recommendation? 
Thanks a lot. I love Pocketbase :)

**Reply by ganigeorgiev** [reply]
> Serverside makes me feel safer.

A lot of users seems to feel that way but I don't understand it. If you use JS SSR you have to manage two different backends and to constantly sync the auth state between them. I don't see how this is more secure.
In any case, for security recommendations about the JS SDK see https://github.com/pocketbase/js-sdk?tab=readme-ov-file#security.
Having a CSP will be enough to mitigate almost all XSS attacks as long as you don't load random 3rd party JS.

> What is your recommendation?

See the footnote in https://github.com/pocketbase/pocketbase/discussions/3363#discussioncomment-7085071 and https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7620032.

**Reply by towbeeWee** [reply]
Thanks Gani for your helpful Tips and super fast replies 👍
I get your points
